File: ch1/zero_matrix/zeroMatrix.cc
#include <vector>
#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <ctime>
using namespace std;
void setUp(vector< vector<int> >& v){
    for(int i = 0; i < v.size(); i++){
        for(int j = 0; j < v[0].size(); j++){
            v[i][j] = rand() % ((v.size() * v[0].size())/5);
        }
    }
}

void printArr(vector< vector<int> >& v){
    for(int i = 0; i < v.size(); i++){
        for(int j = 0; j < v[0].size(); j++){
            printf("%3d", v[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

void zero(vector< vector<int> >& v, int row, int col){
    for(int i = 0; i < v.size(); i++){
        for(int j = 0; j < v[0].size(); j++){
            if(i == row){
                v[i][j] = 0;
            }
            if(j == col){
                v[i][j] = 0;
            }
        }
    }
}

vector< vector<int> > zeroMatrix(vector< vector<int> >& v){
    vector< vector<int> > tmp = v;
    for(int i = 0; i < v.size(); i++){
        for(int j = 0; j < v[0].size(); j++){
            if(v[i][j] == 0){
                zero(tmp, i, j);
            }
        }
    }
    return tmp;
}

int main(int argc, char** argv){
    srand(time(NULL));
    if(argc != 3){
        cout << "Usage: $ ./zeroMatrix col row" << endl;
        return 1;
    } 
    vector< vector<int> > v(stoi(argv[1]), vector<int>(stoi(argv[2])));
    setUp(v);
    printArr(v);
    vector< vector<int> > newV = zeroMatrix(v);
    printArr(newV);
    return 0;
}

File: ch1/zero_matrix/python/zeroMatrix.py
import sys
import numpy as np

# Zero the ith column and jth row
def zero(arr, col, row):
    for i in range(0, len(arr)):
        for j in range(0, len(arr[0])):
            if i == col:
                arr[i][j] = 0
            if j == row:
                arr[i][j] = 0

def zeroMatrix(arr):
    # Copy the array! Or else you zero the whole thing out
    tmp = arr.copy()

    for i in range(0, len(arr)):
        for j in range(0, len(arr[0])):
            if arr[i][j] == 0:
                # I found a zero! now i need to zero out the ith column and jth row
                zero(tmp, i , j)
    # Return new array
    return tmp

# Print the array
def printArr(arr):
    for i in range(0, len(arr)):
        for j in range(0, len(arr[0])):
            print("%3d" % (arr[i][j]), end="")
        print("\n")
    print("\n")

def main():
    # Check for proper Usage
    if(len(sys.argv) != 3):
        print("Usage: $ python zeroMatrix.py col row")
    else:
        # Create a new random array (list)
        begin = np.random.randint((int(sys.argv[1]) * int(sys.argv[2])) / 10, size=(int(sys.argv[1]),int(sys.argv[2])))
        # Print the list before
        printArr(begin)
        # Zero that array out
        newArr = zeroMatrix(begin)
        # Print result
        printArr(newArr)


if __name__ == "__main__":
    main()
